### DISCUSSION TITLE
i am getting TypeError scan carry output and input must have identical types how can i solve this error

### DISCUSSION THREAD

**Discussion by Tanman1234** [jax-ml/jax#6667] [Q&A]
TypeError: scan carry output and input must have identical types, got
ShapedArray(float32[5,1])
and
ShapedArray(float32[5]).

**Comment by jakevdp** [comment]
This probably means that your scan function changed the shape of the carry. Here is a simple example of a scan that produces this error:
```python
from jax import lax
import jax.numpy as jnp

def func(carry, x):
  return (carry[:, None], x)

lax.scan(func, jnp.arange(5.0), jnp.arange(5))
# TypeError: scan carry output and input must have identical types, got
# ShapedArray(float32[5,1])
# and
# ShapedArray(float32[5]).
```
In this case, you can fix the issue by not reshaping the `carry` value:
```python
def func(carry, x):
  return (carry, x)

lax.scan(func, jnp.arange(5.0), jnp.arange(5))
```

**Reply by jakevdp** [reply]
It looks like the `discrete_dynamics()` call within `rollout_looper` returns an array of a different shape than the input, which leads to the error you're seeing.

**Reply by Tanman1234** [reply]
so is there any other way to handle this error or how cani deal with array of different shapes

**Reply by jakevdp** [reply]
A `carry` value must have a constant shape throughout a `scan` operation. I'm not sure what your `discrete_dynamics()` function does, but you may be able to modify it so that it doesn't change the shape of the `carry`.
